test/NuGet.Core.Tests/NuGet.ProjectManagement.Test/MSBuildNuGetProjectTests.cs

```diff
@@ -189,6 +189,237 @@ public class MSBuildNuGetProjectTests
+                Assert.Equal(2, msBuildNuGetProjectSystem.References.Count);
+                Assert.Equal(1, msBuildNuGetProjectSystem.BatchCount);
+                Assert.Equal("a.dll", msBuildNuGetProjectSystem.References.First().Key);
+                Assert.Equal("b.dll", msBuildNuGetProjectSystem.References.Skip(1).First().Key);
+                Assert.Equal(Path.Combine(msBuildNuGetProject.FolderNuGetProject.GetInstalledPath(packageIdentity),
+                    "lib\\net45\\a.dll"), msBuildNuGetProjectSystem.References.First().Value);
+                Assert.Equal(Path.Combine(msBuildNuGetProject.FolderNuGetProject.GetInstalledPath(packageIdentity),
+                    "lib\\net45\\b.dll"), msBuildNuGetProjectSystem.References.Skip(1).First().Value);
+
+                // Main Act
+                await msBuildNuGetProject.UninstallPackageAsync(packageIdentity, testNuGetProjectContext, token);
+                // Check the number of packages and packages returned by PackagesConfigProject after the installation
+                packagesInPackagesConfig = (await msBuildNuGetProject.PackagesConfigNuGetProject.GetInstalledPackagesAsync(token)).ToList();
+                Assert.Equal(0, packagesInPackagesConfig.Count);
+                // Check that the reference has been added to MSBuildNuGetProjectSystem
+                Assert.Equal(0, msBuildNuGetProjectSystem.References.Count);
+                Assert.Equal(2, msBuildNuGetProjectSystem.BatchCount);
+            }
+        }
+
+        [Fact]
+        public async Task TestMSBuildNuGetProjectUninstallReferencesFailureEventBatching()
+        {
+            // Arrange
+            var packageIdentity = new PackageIdentity("packageA", new NuGetVersion("1.0.0"));
+
+            using (var randomTestPackageSourcePath = TestFileSystemUtility.CreateRandomTestFolder())
+            using (var randomPackagesFolderPath = TestFileSystemUtility.CreateRandomTestFolder())
+            using (var randomPackagesConfigFolderPath = TestFileSystemUtility.CreateRandomTestFolder())
+            {
+                var randomPackagesConfigPath = Path.Combine(randomPackagesConfigFolderPath, "packages.config");
+                var token = CancellationToken.None;
+
+                var projectTargetFramework = NuGetFramework.Parse("net45");
+                var testNuGetProjectContext = new TestNuGetProjectContext();
+                var msBuildNuGetProjectSystem = new TestMSBuildNuGetProjectSystem(projectTargetFramework, testNuGetProjectContext);
+                var msBuildNuGetProject = new MSBuildNuGetProject(msBuildNuGetProjectSystem, randomPackagesFolderPath, randomPackagesConfigFolderPath);
+
+                msBuildNuGetProjectSystem.RemoveReferenceAction = (string referenceName) => { throw new InvalidOperationException(); };
+
+                // Pre-Assert
+                // Check that the packages.config file does not exist
+                Assert.False(File.Exists(randomPackagesConfigPath));
```

Comment:
You could probably drop some of these pre-asserts and things not related to the batch. I know the rest of the tests are doing this, but it gets confusing when the tests are testing themselves.